### PR TITLE
Conditional unsimplifier

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/optimiser/ConditionalSimplifier.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+#include <libyul/optimiser/NameCollector.h>
+#include <libdevcore/CommonData.h>
+#include <libdevcore/Visitor.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void ConditionalSimplifier::operator()(Switch& _switch)
+{
```
Comment (Marenz) [inline libyul/optimiser/ConditionalSimplifier.cpp:30]: There seems to be a space at the beginning of each line for this function body

**Comment by leonardoalt** [comment]
Should this one be reviewed instead of https://github.com/ethereum/solidity/pull/7401?

**Comment by chriseth** [comment]
@leonardoalt this one builds on top of #7401.

**Comment by leonardoalt** [comment]
Yes, but the other PR mentions that for it to work it needs the reverse operation.

File: libyul/optimiser/ConditionalSimplifier.cpp (hunk 1 above)
Comment (chriseth) [reply libyul/optimiser/ConditionalSimplifier.cpp:30]: I'm pretty sure it's a tab and not a space. I think github diff just formats it weirdly.

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    if y { revert(0, 0) }
+}
+// ====
+// step: fullSuite
+// ----
+// {
+//     {
+//         let y := mload(0x20)
+//         for { } and(y, 8) { if y { revert(0, 0) } }
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: Why doesn't it see that `y = 0` here?

File: test/libyul/yulOptimizerTests/fullSuite/devcon_example.yul [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// step: fullSuite
+// ----
+// {
+//     {
+//         let _1 := calldataload(0)
+//         let sum := 0
+//         let i := sum
+//         for { } lt(i, calldataload(_1)) { i := add(i, 1) }
+//         {
+//             sum := add(sum, calldataload(add(add(_1, mul(i, 0x20)), 0x20)))
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/fullSuite/devcon_example.yul:27]: Nice!

File: libyul/optimiser/ConditionalUnsimplifier.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+				if (
+					_if.condition->type() == typeid(Identifier) &&
+					!_if.body.statements.empty()
+				)
+				{
+					YulString condition = boost::get<Identifier>(*_if.condition).name;
+					if (
+						_stmt2.type() == typeid(Assignment) &&
+						TerminationFinder(m_dialect).controlFlowKind(_if.body.statements.back()) !=
+						TerminationFinder::ControlFlow::FlowOut
```
Comment (leonardoalt) [inline libyul/optimiser/ConditionalUnsimplifier.cpp]:
```suggestion
						    TerminationFinder::ControlFlow::FlowOut
```

**Review by leonardoalt — COMMENTED** [review]
Looks good, just some minor comments.
Also, can you please add some unit tests for the `ConditionalUnsimplifier`?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `9198073a59502be2095a5c3037db7014949eaf7f`:
```
Coding style error:
 libyul/optimiser/ConditionalUnsimplifier.cpp:81: TerminationFinder::ControlFlow::FlowOut

```
Please check that your changes are working as intended.

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul (hunk 2 above)
Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: It's the full suite, the `y=0` is removed again.

File: test/libyul/yulOptimizerTests/fullSuite/devcon_example.yul (hunk 3 above)
Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/devcon_example.yul:27]: Unfortunately, it only works for `calldataload` and not for `mload`.

**Comment by chriseth** [comment]
Rebased and added tests.

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul (hunk 2 above)
Comment (leonardoalt) [reply test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: Yes, but what I meant is that why doesn't it (some optimizer) see that `y = 0` at post and removes the `if` entirely?

**Review by leonardoalt — APPROVED** [review]

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul (hunk 2 above)
Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: I still do not fully understand. There is `if y {continue}`, so the post-block is executed in both situations `y=0` and `y!=0`.

**Pull request merged**